crates/red_knot_server/src/session.rs

```diff
@@ -0,0 +1,310 @@
+//! Data model, state management, and configuration resolution.
+
+use std::collections::BTreeMap;
+use std::ops::{Deref, DerefMut};
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use lsp_types::{ClientCapabilities, NotebookDocumentCellChange, Url};
+
+use red_knot_workspace::db::RootDatabase;
+use red_knot_workspace::workspace::WorkspaceMetadata;
+use ruff_db::program::{ProgramSettings, SearchPathSettings, TargetVersion};
+use ruff_db::system::SystemPath;
+use ruff_db::Db as _;
+
+use crate::edit::{DocumentKey, DocumentVersion, NotebookDocument};
+use crate::system::LSPSystem;
+use crate::{PositionEncoding, TextDocument};
+
+pub(crate) use self::capabilities::ResolvedClientCapabilities;
+pub use self::index::DocumentQuery;
+pub(crate) use self::settings::AllSettings;
+pub use self::settings::ClientSettings;
+
+mod capabilities;
+pub(crate) mod index;
+mod settings;
+
+/// The global state for the LSP
+pub struct Session {
+    /// Used to retrieve information about open documents and settings.
+    index: Option<Arc<index::Index>>,
+    /// Maps workspace root paths to their respective databases.
+    workspaces: BTreeMap<PathBuf, salsa::Handle<RootDatabase>>,
+    /// The global position encoding, negotiated during LSP initialization.
+    position_encoding: PositionEncoding,
+    /// Tracks what LSP features the client supports and doesn't support.
+    resolved_client_capabilities: Arc<ResolvedClientCapabilities>,
+}
+
+impl Session {
+    pub fn new(
+        client_capabilities: &ClientCapabilities,
+        position_encoding: PositionEncoding,
+        global_settings: ClientSettings,
+        workspace_folders: &[(Url, ClientSettings)],
+    ) -> crate::Result<Self> {
+        let mut workspaces = BTreeMap::new();
+        let index = Arc::new(index::Index::new(workspace_folders, global_settings)?);
+
+        for (url, _) in workspace_folders {
+            let path = url
+                .to_file_path()
+                .map_err(|()| anyhow!("Workspace URL is not a file or directory: {:?}", url))?;
+            let system_path = SystemPath::from_std_path(&path)
+                .ok_or_else(|| anyhow!("Workspace path is not a valid UTF-8 path: {:?}", path))?;
+            let system = LSPSystem::new(index.clone());
```

Comment:
> Do we need a different system for each workspace or could we use a single, shared system for all of them (where `LSPSystem` itself is cheap cloneable?)

Just to confirm, you mean just the system and not the root database as a whole, right?

I think that should be possible, we already use a global index to store the open documents from all workspaces. This would mean that we store the `Index` inside the `LSPSystem` itself and both the `Session` and `RootDatabase` accesses this instead.

Reply:
> It makes me wonder if we could implement `System` on `Index` instead.

This is partly the reason the current solution exists. If we implement `System` on `Index`, we need to use an `Arc` here because it needs to exists on both the `RootDatabase` and `Session`. Otherwise we'll need to find a way to access `Index` from `RootDatabase` (not ideal). And, `Arc` would basically mean that we'd need the `MutIndexGuard`.